File: backend/src/main/java/com/example/webshopbackend/domain/AdvertisedJob.java
package com.example.webshopbackend.domain;

import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "jobs")
public class AdvertisedJob {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Size(
        min = 3,
        max = 25,
        message = "The length of the title should be between 3 and 25 characters"
    )
    private String title;

    @Size(
        max = 1000,
        message = "The length of the description should be maximum 1000 characters"
    )
    private String description;

    @OneToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @JoinColumn(name = "address_id", referencedColumnName = "id")
    private Address address;

    @NotNull
    private String price;

    private JobState state;

    private String imageUrl;

    private String time;

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "users_id", referencedColumnName = "id")
    private User publisher;
}

File: frontend/app/src/main/java/com/example/thesisapp/domain/model/AdvertisedJob.kt
package com.example.thesisapp.domain.model

import com.example.thesisapp.data.disk.model.RoomAdvertisedJob
import com.example.thesisapp.data.network.model.JobResponse

data class AdvertisedJob(
    val id: Long? = null,
    val title: String,
    val description: String,
    val address: Address,
    val coordinates: Coordinates,
    val price: Int,
    val jobState: JobState = JobState.ACTIVE,
    val publisher: User? = null,
    val imageUrl: String,
    val time: String,
    val newMessage: Boolean = false
)

fun JobResponse.toAdvertisedJob() =
    AdvertisedJob(
        id = id,
        title = title,
        description = description,
        address = address,
        coordinates = coordinates,
        price = price,
        jobState = jobState,
        publisher = publisher,
        imageUrl = imageUrl,
        time = time,
        newMessage = newMessage
    )

fun AdvertisedJob.toJobResponse() =
    JobResponse(
        id = id,
        title = title,
        description = description,
        address = address,
        coordinates = coordinates,
        price = price,
        jobState = jobState,
        publisher = publisher,
        imageUrl = imageUrl,
        time = time,
        newMessage = newMessage
    )

fun RoomAdvertisedJob.toAdvertisedJob() =
    AdvertisedJob(
        id = id,
        title = title,
        description = description,
        address = address,
        coordinates = coordinates,
        price = price,
        jobState = jobState,
        publisher = publisher,
        imageUrl = imageUrl,
        time = time,
        newMessage = newMessage
    )

fun AdvertisedJob.toRoomAdvertisedJob() =
    RoomAdvertisedJob(
        id = id,
        title = title,
        description = description,
        address = address,
        coordinates = coordinates,
        price = price,
        jobState = jobState,
        publisher = publisher,
        imageUrl = imageUrl,
        time = time,
        newMessage = newMessage
    )
